From eceab757eb708a8348e0b8d6fa860b26eec5e807 Mon Sep 17 00:00:00 2001
From: Chen Minqiang <ptpt52@gmail.com>
Date: Tue, 20 Apr 2021 11:46:21 +0800
Subject: [PATCH] mt7615: hack to clear all pending packets

---
 mt7615/mcu.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/mt7615/mcu.c b/mt7615/mcu.c
index 856ff36e..2c67abb6 100644
--- a/mt7615/mcu.c
+++ b/mt7615/mcu.c
@@ -750,6 +750,10 @@ mt7615_mcu_add_dev(struct mt7615_phy *phy, struct ieee80211_vif *vif,
 				 &data, sizeof(data), true);
 }
 
+static int
+__mt7615_mcu_add_sta(struct mt76_phy *phy, struct ieee80211_vif *vif,
+		     struct ieee80211_sta *sta, bool enable, int cmd);
+
 static int
 mt7615_mcu_add_beacon_offload(struct mt7615_dev *dev,
 			      struct ieee80211_hw *hw,
@@ -815,6 +819,11 @@ mt7615_mcu_add_beacon_offload(struct mt7615_dev *dev,
 	}
 	dev_kfree_skb(skb);
 
+	if (is_mt7663(&dev->mt76)) {
+		//hack to clear all pending packets
+		__mt7615_mcu_add_sta(dev->phy.mt76, vif, NULL, false, MCU_EXT_CMD_STA_REC_UPDATE);
+		__mt7615_mcu_add_sta(dev->phy.mt76, vif, NULL, true, MCU_EXT_CMD_STA_REC_UPDATE);
+	}
 out:
 	return mt76_mcu_send_msg(&dev->mt76, MCU_EXT_CMD_BCN_OFFLOAD, &req,
 				 sizeof(req), true);
@@ -1191,6 +1200,12 @@ static int
 mt7615_mcu_add_sta(struct mt7615_phy *phy, struct ieee80211_vif *vif,
 		   struct ieee80211_sta *sta, bool enable)
 {
+	if (is_mt7663(&phy->dev->mt76) && !enable) {
+		//hack to clear all pending packets
+		__mt7615_mcu_add_sta(phy->mt76, vif, sta, enable, MCU_EXT_CMD_STA_REC_UPDATE);
+		__mt7615_mcu_add_sta(phy->mt76, vif, NULL, false, MCU_EXT_CMD_STA_REC_UPDATE);
+		__mt7615_mcu_add_sta(phy->mt76, vif, NULL, true, MCU_EXT_CMD_STA_REC_UPDATE);
+	}
 	return __mt7615_mcu_add_sta(phy->mt76, vif, sta, enable,
 				    MCU_EXT_CMD_STA_REC_UPDATE);
 }
